File: forms/medical-operation-note/front-end-form-with-svelte/src/lib/engine/never-event-rules.ts
import type { CompositeRisk, FiredRule, OperationNote } from './types.js';

/**
 * NHS England Never Events (2018 policy) relevant to the operating
 * theatre, parsed from a comma-separated list on
 * `safetyCountsEbl.neverEventFlags`.
 *
 * Accepted tokens (case-insensitive, trimmed):
 *   wrong-site, wrong-side, wrong-patient, wrong-procedure, wrong-implant
 *
 * Plus the standalone Boolean `retainedForeignBody === 'yes'`.
 *
 * Any of these forces composite risk = Critical.
 */
const KNOWN_TOKENS = new Set([
  'wrong-site',
  'wrong-side',
  'wrong-patient',
  'wrong-procedure',
  'wrong-implant',
]);

export function applyNeverEventRules(data: OperationNote): {
  fired: boolean;
  risk: CompositeRisk;
  firedRules: FiredRule[];
  tokens: string[];
  retainedForeignBody: boolean;
} {
  const firedRules: FiredRule[] = [];
  const raw = data.safetyCountsEbl.neverEventFlags || '';
  const tokens = raw
    .split(',')
    .map((t) => t.trim().toLowerCase())
    .filter((t) => t.length > 0 && KNOWN_TOKENS.has(t));

  for (const t of tokens) {
    firedRules.push({
      ruleId: `R-NE-${t.toUpperCase()}`,
      instrument: 'never-event',
      grade: 'critical',
      category: 'never-event',
      description: `Never-event suspicion: ${t}`,
    });
  }

  const retainedForeignBody = data.safetyCountsEbl.retainedForeignBody === 'yes';
  if (retainedForeignBody) {
    firedRules.push({
      ruleId: 'R-NE-RETAINED-FOREIGN-BODY',
      instrument: 'never-event',
      grade: 'critical',
      category: 'never-event',
      description: 'Retained foreign body declared',
    });
  }

  if (data.safetyCountsEbl.intraOperativeArrest === 'yes') {
    firedRules.push({
      ruleId: 'R-NE-INTRA-OP-ARREST',
      instrument: 'never-event',
      grade: 'critical',
      category: 'arrest',
      description: 'Intra-operative cardiac or respiratory arrest',
    });
  }

  const fired = tokens.length > 0 || retainedForeignBody ||
    data.safetyCountsEbl.intraOperativeArrest === 'yes';

  return {
    fired,
    risk: fired ? 'critical' : 'routine',
    firedRules,
    tokens,
    retainedForeignBody,
  };
}

File: forms/medical-operation-note/full-stack-with-loco-tera-htmx-alpine/medical_operation_note/src/engine/never_event_rules.rs
//! Never-event rules (NHS England Never Events Policy & Framework).
//!
//! Any never-event candidate (wrong site / side / patient / procedure /
//! implant, retained foreign object, intra-operative arrest) escalates the
//! composite risk to Critical and fires a high-priority flag.

use super::types::{CompositeRisk, FiredRule};

pub fn evaluate(
    never_event_flagged: bool,
    never_event_kind: &str,
    intra_operative_arrest: bool,
) -> (CompositeRisk, Vec<FiredRule>) {
    let mut rules: Vec<FiredRule> = Vec::new();
    let mut risk = CompositeRisk::Routine;

    if never_event_flagged {
        risk = risk.max(CompositeRisk::Critical);
        rules.push(FiredRule {
            rule_id: "never-event.flagged".into(),
            instrument: "never-event".into(),
            band: "critical".into(),
            category: "never-event".into(),
            description: format!(
                "Never-event candidate flagged{}",
                if never_event_kind.is_empty() {
                    String::new()
                } else {
                    format!(": {never_event_kind}")
                }
            ),
        });
    }
    if intra_operative_arrest {
        risk = risk.max(CompositeRisk::Critical);
        rules.push(FiredRule {
            rule_id: "arrest.intraop".into(),
            instrument: "anaesthetic".into(),
            band: "critical".into(),
            category: "intra-operative-arrest".into(),
            description: "Cardiac or respiratory arrest in theatre".into(),
        });
    }

    (risk, rules)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn nothing_routine() {
        let (r, rules) = evaluate(false, "", false);
        assert_eq!(r, CompositeRisk::Routine);
        assert!(rules.is_empty());
    }

    #[test]
    fn flagged_critical() {
        let (r, rules) = evaluate(true, "wrong-site", false);
        assert_eq!(r, CompositeRisk::Critical);
        assert_eq!(rules.len(), 1);
        assert!(rules[0].description.contains("wrong-site"));
    }

    #[test]
    fn arrest_critical() {
        let (r, rules) = evaluate(false, "", true);
        assert_eq!(r, CompositeRisk::Critical);
        assert_eq!(rules.len(), 1);
    }
}
